packages/babel-parser/src/parser/expression.js

```diff
@@ -1934,14 +1934,33 @@ export default class ExpressionParser extends LValParser {
       const oldLabels = this.state.labels;
       this.state.labels = [];
       if (useStrict) this.state.strict = true;
-      // Add the params to varDeclaredNames to ensure that an error is thrown
-      // if a let/const declaration in the function clashes with one of the params.
-      this.checkParams(
-        node,
-        !oldStrict && !useStrict && !allowExpression && !isMethod && !nonSimple,
-        allowExpression,
-      );
-      node.body = this.parseBlock(true, false);
+      node.body = this.parseBlock(true, false, () => {
+        if (this.state.strict && nonSimple) {
```

Comment:
Note that the previous implementation only threw if there is an `"use strict"` directive inside the current function, not if the function is in strict mode.

This makes me realize that my suggestion wasn't 100% correct: we need to pass a parameter `hasStrictDirective` to this function, which is `true` if the `parseBlockOrModuleBlockBody` found a `"use strict"` directive.